### PR TITLE
Validate that all context objects specified in labels are referred to…

### PR DESCRIPTION
**Pull request by qchaupds**
… in the parent/bundle collection reference list

1. Add new class to hold information about a reference: src/main/java/gov/nasa/pds/tools/util/HashSetReferenceInfo.java
2. Add new class to refactor code: src/main/java/gov/nasa/pds/tools/util/ReferentialIntegrityUtil.java
3. Add new test resources: src/test/resources/github69
4. Modify some fields and functions to be public: src/main/java/gov/nasa/pds/tools/util/LabelUtil.java
5. Add new type: src/main/java/gov/nasa/pds/tools/validate/ProblemType.java
6. Move code to refactor class ReferentialIntegrityUtil: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/BundleReferentialIntegrityRule.java
7. Modify to use refactor function: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/CollectionReferentialIntegrityRule.java
8. Add new flag skip-context-reference-check and initialize/reset ReferentialIntegrityUtil class: src/main/java/gov/nasa/pds/validate/ValidateLauncher.java
9. Add new flag skip-context-reference-check: src/main/java/gov/nasa/pds/validate/commandline/options/ConfigKey.java
10. Add new flag skip-context-reference-check: src/main/java/gov/nasa/pds/validate/commandline/options/Flag.java
11. Add new flag skip-context-reference-check: src/main/java/gov/nasa/pds/validate/commandline/options/FlagOptions.java
12. Add new problem types warning.integrity.missing_context_reference: src/test/java/cucumber/StepDefs.java
13. Modify tests due to new warning messsage and add 3 new tests for new issue: src/test/resources/features/validate.feature

Refs #69 https://github.com/NASA-PDS/validate/issues/69 As a user, I want to validate that all context objects specified in observational products are referenced in the parent bundle/collection Reference_List

Validate that all context objects specified in label and referred to in the parent/bundle collection reference list

Closes NASA-PDS/validate/#69

Tested in DEV with:

% validate -R pds4.bundle --skip-product-validation -r report_github69_bundle_invalid_1.json   -s json -t src/test/resources/github69/invalid/bundle_kaguya_derived.xml >& t2

The reported warning messages for parent bundle is because the ID specified in the collection is not referenced in the parent bundle:

      "label": "file:/home/qchau/sandbox/validate/src/test/resources/github69/invalid/data_spectra/spectra_data_collection_inventory.xml",
          "message": "The context reference 'urn:nasa:pds:context:investigation:mission.kaguya' could not be found in the parent bundle reference list. (Disable with --skip-context-reference-check flag)"

{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 210 % grep -rn urn:nasa:pds:context:investigation:mission.kaguya src/test/resources/github69/invalid | grep collection
src/test/resources/github69/invalid/data_spectra/spectra_data_collection_inventory.xml:45:        <lid_reference>urn:nasa:pds:context:investigation:mission.kaguya</lid_reference>

      "label": "file:/home/qchau/sandbox/validate/src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per1.xml",
          "message": "The context reference 'urn:nasa:pds:context:instrument_host:spacecraft.kaguya' could not be found in the parent collection reference list. (Disable with --skip-context-reference-check flag)"
          "message": "The context reference 'urn:nasa:pds:context:instrument:kaguya.grs' could not be found in the parent collection reference list. (Disable with --skip-context-reference-check flag)"
          "message": "The context reference 'urn:nasa:pds:kaguya_grs_spectra:document:kgrs_calibrated_spectra' could not be found in the parent collection reference list. (Disable with --skip-context-reference-check flag)"

The above references are declared in the labels but not found in the parent collection.

{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 213 % grep -rn urn:nasa:pds:context:instrument:kaguya.grs  src/test/resources/github69/invalid/data_spectra/
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per1.xml:82:              <lid_reference>urn:nasa:pds:context:instrument:kaguya.grs</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per2.xml:86:              <lid_reference>urn:nasa:pds:context:instrument:kaguya.grs</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per3.xml:86:              <lid_reference>urn:nasa:pds:context:instrument:kaguya.grs</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per4.xml:86:              <lid_reference>urn:nasa:pds:context:instrument:kaguya.grs</lid_reference>
{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 214 % grep -rn urn:nasa:pds:context:instrument_host:spacecraft.kaguya  src/test/resources/github69/invalid/data_spectra/
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per1.xml:74:              <lid_reference>urn:nasa:pds:context:instrument_host:spacecraft.kaguya</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per2.xml:78:              <lid_reference>urn:nasa:pds:context:instrument_host:spacecraft.kaguya</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per3.xml:78:              <lid_reference>urn:nasa:pds:context:instrument_host:spacecraft.kaguya</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per4.xml:78:              <lid_reference>urn:nasa:pds:context:instrument_host:spacecraft.kaguya</lid_reference>
{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 215 % grep -rn urn:nasa:pds:context:instrument:kaguya.grs  src/test/resources/github69/invalid/data_spectra/
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per1.xml:82:              <lid_reference>urn:nasa:pds:context:instrument:kaguya.grs</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per2.xml:86:              <lid_reference>urn:nasa:pds:context:instrument:kaguya.grs</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per3.xml:86:              <lid_reference>urn:nasa:pds:context:instrument:kaguya.grs</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per4.xml:86:              <lid_reference>urn:nasa:pds:context:instrument:kaguya.grs</lid_reference>
{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 216 % grep -rn urn:nasa:pds:kaguya_grs_spectra:document:kgrs_calibrated_spectra  src/test/resources/github69/invalid/data_spectra/
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per1.xml:94:                   <lid_reference>urn:nasa:pds:kaguya_grs_spectra:document:kgrs_calibrated_spectra</lid_reference> 
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per2.xml:102:                  <lid_reference>urn:nasa:pds:kaguya_grs_spectra:document:kgrs_calibrated_spectra</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per3.xml:102:                  <lid_reference>urn:nasa:pds:kaguya_grs_spectra:document:kgrs_calibrated_spectra</lid_reference>
src/test/resources/github69/invalid/data_spectra/kgrs_calibrated_spectra_per4.xml:102:                  <lid_reference>urn:nasa:pds:kaguya_grs_spectra:document:kgrs_calibrated_spectra</lid_reference>

% validate -R pds4.bundle --skip-product-validation -r report_github69_bundle_invalid_2.json   -s json -t src/test/resources/github69/invalid_2/bundle_kaguya_derived.xml >& t3

If the flag  --skip-context-reference-check is used, no warning messages regarding missing_context_reference are reported.

% validate -R pds4.bundle --skip-product-validation --skip-context-reference-check -r report_github69_bundle_invalid_3.json   -s json -t src/test/resources/github69/invalid_2/bundle_kaguya_derived.xml

### CODE REVIEW

**Review by jordanpadams — CHANGES_REQUESTED** [review]
@qchaupds the bundle/collection validation appears to be working as expected, but the context object checks do not.

So I ran this against the insight document bundle: https://pds.nasa.gov/data/pds4/test-data/registry/urn-nasa-pds-insight_documents/ (this is out on the dev machines already)

And I got this WARNING:
```
  PASS: file:/data/pds4/test-data/registry/urn-nasa-pds-insight_documents/document_seis/collection_document_seis.xml
      WARNING  [warning.integrity.missing_context_reference]   The context reference 'urn:nasa:pds:insight_documents:document_seis:seis_release_notes' could not be found in the parent bundle reference list. (Disable with --skip-context-reference-check flag)
```

But `urn:nasa:pds:insight_documents:document_seis:seis_release_notes` is not a context object. I may have mispoken over some of our emails. When we parse all of these products for these context objects, we should only be looking within the `/Context_Area` class for the set of `lid[vid]_reference`s. The `Reference_List` (or elsewhere) may contain additional references, but that is a corner case I don't think we need to tackle right now. Sorry for any confusion.

**Review by jordanpadams — CHANGES_REQUESTED** [review]
@qchaupds additionally, I would prefer the WARNING about the missing context objects be with the object missing them. In other words 

instead of the following:
```
  PASS: file:/bundle/collection.xml
      WARNING  [warning.integrity.missing_context_reference]   The context reference 'my context reference' could not be found in the parent bundle reference list. (Disable with --skip-context-reference-check flag)
```

I think it would be more useful for users if they appears with the parent bundle:
```
  PASS: file:/bundle.xml
      WARNING  [warning.integrity.missing_context_reference]   The context reference 'my.context.object.A' could not be found in this bundle, but it was defined in `PRODUCT_Y`. (Disable with --skip-context-reference-check flag)
      WARNING  [warning.integrity.missing_context_reference]   The context reference 'my.context.object.B' could not be found in this bundle, but it was defined in `PRODUCT_Y`. (Disable with --skip-context-reference-check flag)
```
If it is too hard to track `PRODUCT_Y`'s LIDVID, then we can just say `bundle's child product`.

File: src/main/java/gov/nasa/pds/tools/util/ReferentialIntegrityUtil.java [hunk 1]
```diff
@@ -0,0 +1,951 @@
...
+                labelIsBundleFlag = true;
+                // Save the URL of the bundle to be used to report the error.
+                ReferentialIntegrityUtil.parentBundleURL = url;
+            }
+            matcherBundleCollection = COLLECTION_LABEL_PATTERN.matcher(FilenameUtils.getName(child.toString()));
+            if (matcherBundleCollection.matches()) {
+                labelIsCollectionFlag = true;
+            }
+
+            xml = db.parse(url.openStream());
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/util/ReferentialIntegrityUtil.java:865]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/ReferentialIntegrityUtil.java [hunk 2]
```diff
@@ -0,0 +1,951 @@
...
+    DocumentBuilder db = null;
+    Document xml = null;
+    DOMSource domSource = null;
+
+    boolean labelIsCollectionFlag = false;
+    boolean labelIsBundleFlag = false;
+    String parentId = null;
+
+    try {
+      List<Target> children = getContext().getCrawler().crawl(crawlTarget,true);  // Get also the directories.
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/util/ReferentialIntegrityUtil.java]: @qchaupds can we take a look at these unresolved @sonatype-lift comments and look to resolve them?

[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/ReferentialIntegrityUtil.java]: I resolved them after the initial commit:

https://github.com/NASA-PDS/validate/pull/363/commits

https://github.com/NASA-PDS/validate/commit/0a9ef9f93fe1ba0f3487b38634655d543b9e1405

[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/ReferentialIntegrityUtil.java]: @jordanpadams See reply above (forgot the at symbol)

**Review by jordanpadams — APPROVED** [review]

**Pull request merged**